### ISSUE TITLE
Incorrect Reporting of reportOptionalMemberAccess

### ISSUE THREAD

**Issue by rosen-dash** [microsoft/pyright#9188]
```py
class Base:
    def __init__(self, client: dict[str, str]):
        self._client = None

        if client:
            self._client = client
        else:
            self._client = {"k1": "v1"}

    def test(self):
        self._client.get("x")
```

Gives this error

`optional_member_access_test.py:13:22 - error: "get" is not a known attribute of "None" (reportOptionalMemberAccess)`

**Comment by erictraut** [comment]
Pyright is working correctly here.

If you do not provide a type declaration for an instance variable, pyright infers its type from all assignments to the instance variable within the class. 

If your intent is for `_client` to be restricted to the type `dict[str, str]`, you should delete the initial assignment of `None`.

**Comment by rosen-dash** [comment]
> Pyright is working correctly here.
> 
> If you do not provide a type declaration for an instance variable, pyright infers its type from all assignments to the instance variable within the class.
> 
> If your intent is for `_client` to be restricted to the type `dict[str, str]`, you should delete the initial assignment of `None`.

Thank you @erictraut for the quick response. However, this error doesn't come up if I try to access it within `__init__` itself. The order of assignment is honoured within the same function. I have updated the code snippet reflecting the same.

**Comment by erictraut** [comment]
Yes, that's because the expression `self._client` is locally narrowed within the `__init__` method. Type narrowing doesn't apply across execution contexts (methods) because the order in which functions or methods is called cannot be determined statically.

Refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing) for more details about how type narrowing works.

**Comment by rosen-dash** [comment]
Thank you for the taking time to explain and the reference.

> the order in which functions or methods is called cannot be determined statically

Does this hold true for `__init__`? Is it not guaranteed to be called just after the object construction (__new__) and before any other instance method execution?
